lang/LangPrimSource/OSCData.cpp

```diff
@@ -652,6 +681,55 @@ static PyrObject* ConvertOSCMessage(int inSize, char* inData) {
     return obj;
 }
 
+// Convert raw OSC bundle to Array ([time, elements...]).
+static PyrObject* ConvertOSCBundle(int inSize, const char* inData) {
+    int64 oscTime = OSCtime(inData + 8);
+    const char* data = inData + 16;
+    const char* dataEnd = inData + inSize;
+
+    // first count all elements
+    int numElements = 0;
+    for (const char* ptr = data; ptr < dataEnd;) {
+        int32 size = OSCint(ptr);
+        if (size > 0) {
+            ptr += sizeof(int32) + size;
+            numElements++;
+        } else {
+            throw std::runtime_error("Bad OSC bundle element size");
+        }
+    }
+
+    VMGlobals* g = gMainVMGlobals;
+    PyrObject* result = newPyrArray(g->gc, numElements + 1, 0, false);
+    PyrSlot* slots = result->slots;
+
+    if (oscTime != 1) {
+        double seconds = (double)oscTime * kOSCtoSecs;
+        SetFloat(slots, seconds);
+    } else {
+        SetNil(slots); // immediate
+    }
+
+    for (int i = 0; i < numElements; i++) {
+        int32 size = OSCint(data);
+        data += sizeof(int32);
+        PyrObject* bundleElement;
+        if (IsBundle(data)) {
+            bundleElement = ConvertOSCBundle(size, data);
+        } else if (IsMessage(data)) {
+            bundleElement = ConvertOSCMessage(size, data);
+        } else {
+            throw std::runtime_error("Malformed OSC bundle element");
+        }
+        SetObject(slots + i + 1, (PyrObjectHdr*)bundleElement);
```

Comment:
Is this cast needed? Can't you implicitly convert to a base class pointer? If not, why not a static cast?

Reply:
> Can't you implicitly convert to a base class pointer?

Good point! I probably didn't realize that `PyrObject` *inherits* from `PyrObjHdr`... Will remove the cast.